rig/place_and_route/utils.py

```diff
@@ -4,13 +4,187 @@
+        chip. This resource will be set to the number of bytes in the largest
+        free block in the SRAM heap. This gives a conservative estimate of the
+        amount of free SRAM on the chip which will be an underestimate in the
+        presence of memory fragmentation.
+
+    Returns
+    -------
+    :py:class:`rig.machine.Machine`
+        A :py:class:`~rig.machine.Machine` object representing the
+        resources available within a SpiNNaker machine in the form used by the
+        place-and-route infrastructure.
+    """
+    try:
+        max_cores = max(c.num_cores for c in itervalues(system_info))
+    except ValueError:
+        max_cores = 0
+
+    try:
+        max_sdram = max(c.largest_free_sdram_block
+                        for c in itervalues(system_info))
+    except ValueError:
+        max_sdram = 0
+
+    try:
+        max_sram = max(c.largest_free_sram_block
+                       for c in itervalues(system_info))
+    except ValueError:
+        max_sram = 0
+
+    return Machine(width=system_info.width,
+                   height=system_info.height,
+                   chip_resources={
+                       core_resource: max_cores,
+                       sdram_resource: max_sdram,
+                       sram_resource: max_sram,
+                   },
+                   chip_resource_exceptions={
+                       chip: {
+                           core_resource: info.num_cores,
+                           sdram_resource: info.largest_free_sdram_block,
+                           sram_resource: info.largest_free_sram_block,
+                       }
+                       for chip, info in iteritems(system_info)
+                       if (info.num_cores != max_cores
+                           or info.largest_free_sdram_block != max_sdram
+                           or info.largest_free_sram_block != max_sram)
+                   },
+                   dead_chips=set(system_info.dead_chips()),
+                   dead_links=set(system_info.dead_links()))
+
+
+def build_core_constraints(system_info, machine, core_resource=Cores):
+    """Return a set of place-and-route
+    :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint`
+    which reserve any cores that that are already in use.
+
+    The returned list of
+    :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint`\ s
+    reserves all cores not in an Idle state (i.e. not a monitor and not already
+    running an application).
+
+    .. note::
+
+        Historically it was required that a
+        :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint`
+        was manually added by every application which reserved the monitor core
+        on each chip. Improves on this by automatically generating appropriate
+        constraints for reserving not just monitor cores but also other cores
+        which are already in use.
```

Comment:
"Historically, every application was required to add a :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint` to reserve the monitor processor on each chip. This method improves upon this approach by automatically generating constraints which reserve not just the monitor core but also any other cores which are already in use."